pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -242,41 +243,68 @@ func (n *NMStateConfig) validateNMStateLabels() field.ErrorList {
 	return allErrs
 }
 
-func getFirstIP(nmStateConfig *nmStateConfig) string {
+func getFirstIP(nmstateRaw []byte) (string, error) {
+	var nmStateConfig nmStateConfig
+	err := yaml.Unmarshal(nmstateRaw, &nmStateConfig)
+	if err != nil {
+		return "", fmt.Errorf("error unmarshalling NMStateConfig: %w", err)
+	}
+
 	for _, intf := range nmStateConfig.Interfaces {
 		for _, addr4 := range intf.IPV4.Address {
 			if addr4.IP != "" {
-				return addr4.IP
+				return addr4.IP, nil
 			}
 		}
 		for _, addr6 := range intf.IPV6.Address {
 			if addr6.IP != "" {
-				return addr6.IP
+				return addr6.IP, nil
 			}
 		}
 	}
-	return ""
+
+	return "", nil
 }
 
-// GetNodeZeroIP retrieves the first IP from the user provided NMStateConfigs to set as the node0 IP
-func GetNodeZeroIP(nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
-	for i := range nmStateConfigs {
-		var nmStateConfig nmStateConfig
-		err := yaml.Unmarshal(nmStateConfigs[i].Spec.NetConfig.Raw, &nmStateConfig)
+// GetNodeZeroIP retrieves the first IP to be set as the node0 IP.
+// The method prioritizes the search by trying to scan first the NMState configs defined
+// in the agent-config hosts - so that it would be possible to skip the worker nodes - and then
+// the NMStateConfig.
+func GetNodeZeroIP(agentConfig *agenttypes.Config, nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
+	rawConfigs := []aiv1beta1.RawNetConfig{}
+
+	// Select first the configs from the hosts, if defined
+	if agentConfig != nil {
+		for _, host := range agentConfig.Hosts {
+			// Skip worker hosts (or without an explicit role assigned)
+			if host.Role != "master" {
+				continue
+			}
+			rawConfigs = append(rawConfigs, host.NetworkConfig.Raw)
+		}
+	}
```

Comment:
This could still end up selecting a worker (due to falling back to the nmstateconfigs below) in the case [`"worker"`, `"worker"`, `""`,  `""`, `""`].

This could be solved by adding a second loop after this:

```
if agentConfig != nil {
		for _, host := range agentConfig.Hosts {
			// Skip worker hosts
			if host.Role != "" {
				continue
			}
			rawConfigs = append(rawConfigs, host.NetworkConfig.Raw)
		}
	}
```

That would cover almost all cases - everything except the case where the workers have implicit role assignment, but only workers have network configs with static IPs. Complete correctness could be achieved by doing this conditionally only when we didn't already find enough explicit masters to equal the number of control plane replicas.

Reply:
Yeah I think it's working (I've added another test for that) (assuming you meant except for the case with the implict masters). I'm still even more convinced that if we had opted (initially) for mandatory roles the overall logic could have been simpler, also in the validations.